### PR TITLE
Fix qpy for `MCX` gates

### LINKED ISSUE #9390: QPY fails to serialize `MCXGrayCode` gate
### Environment

- **Qiskit Terra version**: 0.22.4
- **Python version**: 3.9
- **Operating system**: macOS Monterey 12.6

### What is happening?

When you try to serialize a circuit with an `MCXGrayCode` gate using qpy, you get `TypeError: __init__() missing 1 required positional argument: 'num_ctrl_qubits'`. As far as I have seen, other mcx gates do not run into this issue.

I discovered this while trying to run iterative amplitude estimation using the runtime `Sampler`, which relies on `qpy` for serialization.

### How can we reproduce the issue?

```
  from qiskit import QuantumCircuit, qpy
  circuit = QuantumCircuit(11)
  circuit.mcx(list(range(0,9)), 10)
  print('gate name:', circuit.data[0][0].name)

  with open('test.qpy', 'wb') as fd:
      qpy.dump(circuit, fd)

  with open('test.qpy', 'rb') as fd:
      new_qc = qpy.load(fd)[0]
```
This snippet will fail with a similar traceback to:

```
gate name: mcx_gray
  File "/Users/ept/Downloads/ExamplePEO.py", line 131, in <module>
    new_qc = qpy.load(fd)[0]
  File "/Users/ept/Desktop/qgt/qiskit/qpy/interface.py", line 269, in load
    loader(
  File "/Users/ept/Desktop/qgt/qiskit/qpy/binary_io/circuits.py", line 864, in read_circuit
    _read_instruction(file_obj, circ, out_registers, custom_operations, version, vectors)
  File "/Users/ept/Desktop/qgt/qiskit/qpy/binary_io/circuits.py", line 258, in _read_instruction
    gate = gate_class(*params)
TypeError: __init__() missing 1 required positional argument: 'num_ctrl_qubits'
```

### What should happen?

It should not have failed.

### Any suggestions?

I would modify [the following line](https://github.com/Qiskit/qiskit-terra/blob/main/qiskit/qpy/binary_io/circuits.py#L255) of `qpy.binary_io.circuits.py` to add "MCXGrayCode". There seem to be other gates that have been already accounted for this same issue, and adding "MCXGrayCode" to the list fixes it:

```
        if gate_name in {"MCPhaseGate", "MCU1Gate", "MCXGrayCode"}:
            gate = gate_class(*params, instruction.num_ctrl_qubits)
```

I'm happy to open a bugfix PR, but because I'm not that familiar with qpy, I wanted to document the problem in this issue first.

### PR DESCRIPTION
**Pull request by ElePT**
### Summary
This PR fixes #9390.

### CODE REVIEW

**Review by jakelishman — COMMENTED** [review]
This fix looks fine to me, thanks. You can add a use of this gate into this existing test: https://github.com/Qiskit/qiskit-terra/blob/a0466b36c7086166f780747d501b0fbcdc2863cd/test/python/circuit/test_circuit_load_from_qpy.py#L1044

The tests of QPY are in `test/python/circuit` because that's where QPY itself used to live, before it got its own subpackage. In the same file there are also some larger tests of things like QFT - if there's an intermediate algorithm library object for IPE, you could add a similar test to the QFT one too.

This wants a bugfix release note too.

**Review by usesnames — APPROVED** [review]

**Review by jakelishman — DISMISSED** [review]
Looks good to me, but we'll wait until everything else for 0.23.0 is merged before we merge this, to avoid holding up the release with CI.

This can get merged between rc1 and the full release, since it's a pure bugfix.

**Comment by ElePT** [comment]
Thanks @jakelishman!! That sounds good.

File: qiskit/qpy/binary_io/circuits.py
Comment (mtreinish) [inline qiskit/qpy/binary_io/circuits.py:255]: I'm just curious if there are other gates in the mcx family that have the same constructor and need this. I always get lost in all the mcx variants. But it might be good to check we've got all of them covered here

Comment (ElePT) [reply qiskit/qpy/binary_io/circuits.py:255]: You were right @mtreinish, qpy also fails for other MCX gates. Namely (I think these are all):
```
MCXRecursive
MCXVChain
MCXGate
```
I can add these to the PR. I though I had tested it before realizing that with 2 control qubits, the gate ends up being of type `CCXGate`, so of course all my checks passed.

**Review by jakelishman — APPROVED** [review]

File: test/python/circuit/test_circuit_load_from_qpy.py
Comment (ElePT) [inline test/python/circuit/test_circuit_load_from_qpy.py:1067]:
```suggestion
        qc = QuantumCircuit(6)
        mcu1_gate = MCU1Gate(np.pi, 2)
        mcx_gate = MCXGate(5)
        mcx_gray_gate = MCXGrayCode(5)
        mcx_recursive_gate = MCXRecursive(4)
        mcx_vchain_gate = MCXVChain(3)
        qc.append(mcu1_gate, [0, 2, 1])
        qc.append(mcx_gate, list(range(0, 6)))
        qc.append(mcx_gray_gate, list(range(0, 6)))
        qc.append(mcx_recursive_gate, list(range(0, 5)))
        qc.append(mcx_vchain_gate, list(range(0, 5)))
```

**Review by jakelishman — APPROVED** [review]

File: test/python/circuit/test_circuit_load_from_qpy.py
Comment (mtreinish) [inline test/python/circuit/test_circuit_load_from_qpy.py:1057]:
```suggestion
        qc = QuantumCircuit(9)
```

File: test/python/circuit/test_circuit_load_from_qpy.py
Comment (mtreinish) [inline test/python/circuit/test_circuit_load_from_qpy.py]: It looks like `MCXRecursive` also has an ancilla qubit in addition to n control qubits.
```suggestion
        qc.append(mcx_recursive_gate, list(range(7)))
```

File: test/python/circuit/test_circuit_load_from_qpy.py
Comment (mtreinish) [inline test/python/circuit/test_circuit_load_from_qpy.py]: It looks like `MCXVChain` has 3 ancillas in addition to the n control qubits

```suggestion
        qc.append(mcx_vchain_gate, list(range(9)))
```

**Comment by mtreinish** [comment]
Oh, I was 2min too slow on my comments just ignore those suggestions, decreasing the control qubit counts on the gates works just as well

**Pull request merged**
